aodndata/aatams_sattag_qc_dm/aatams_sattag_qc_dm_schema.py

```diff
@@ -0,0 +1,685 @@
+import os
+from datetime import datetime
+from functools import partial
+
+from schema import And, Or, Schema, Use
+
+from .csv_schema import CSVSchema
+
+AATAMS_DM_NUMBER_OF_FILES_IN_ZIP = 7
+AATAMS_DM_FILE_TYPE_NAMES = (
+    "metadata",
+    "ctd",
+    "diag",
+    "dive",
+    "haulout",
+    "ssmoutputs",
+    "summary",
+)
+
+DATE_FMT_STR_1 = "%Y-%m-%dT%H:%M:%SZ"
+DATE_FMT_STR_2 = "%m/%d/%y %H:%M:%S"
+
+
+# schema functions -> Return arg or False
+def valid_device(astr):  # TODO expand to particular devices fmts?
+    """Check if a valid device.
+
+    Args:
+      astr: a device string
+
+    Returns:
+      str: astr if valid
+      bool: False if invalid
+
+    """
+    return astr if len(astr.split("-")) >= 3 else False
+
+
+def valid_wmo_device(astr):  # TODO expand to particular codes?
+    """Check if a valid wmo device.
+
+    Args:
+      astr: a wmo device string
+
+    Returns:
+      str: astr if valid
+      bool: False if invalid
+
+    """
+    return astr if astr[0] == "Q" else False
+
+
+def valid_tag_type(astr):  # TODO expand to particular strings.
+    """Check if a valid tag type.
+
+    Args:
+      astr: a tag type string
+
+    Returns:
+      str: astr if valid
+      bool: False if invalid
+
+    """
+    return astr if "_" in astr else False
+
+
+def valid_longitude(afloat):  # TODO check convention
+    """Check if a valid longitude.
+
+    Args:
+      afloat: a longitude number
+
+    Returns:
+      float: afloat if valid
+      bool: False if invalid
+
+    """
+    valid = -180.0 <= afloat <= 180.0
+    return afloat if valid else False
+
+
+def valid_latitude(afloat):
+    """Check if a valid latitude.
+
+    Args:
+      afloat: a latitude number
+
+    Returns:
+      float: afloat if valid
+      bool: False if invalid
+
+    """
+    valid = -90 <= afloat <= 90.0
+    return afloat if valid else False
+
+
+def not_applicable(astr):  # TODO: handle every case as particular?
+    """Check string for emptiness, not applicable or NaN string.
+
+    Args:
+      astr: a string
+
+    Returns:
+      str: "" if valid
+      bool: False if invalid
+
+    """
+    if astr in ("", "NA", "NaN"):
```

Comment:
Funny thing - some comments arent in the github review window only in the PR window - completely missed those.

I didn't use pandas because to import it just to do that is overkill. Also, to type check the result of , say, ```pd.to_numeric```, I would need to check against numpy types instead of built-ins, which would increase the number of possible definitions (check for float, check for np.float64, np.float32, etc).

Finally, I prefer to keep the result as much as close to the actual data being read (a string) - that's why I put an empty string.

Yes - I'll add a check for white spaces and the UPPERcase cases.

Reply:
> Yes - I'll add a check for white spaces and the UPPERcase cases.

I can't see this added in the function. Is it somewhere else?

Reply:
done - see fd82b0f8ec2234c15b1551bea7b0b7f3ffe7fc35